'Custom Yggdrasil' AccountType

evan-goode opened the pull request:
Copying this pull request over from https://github.com/PolyMC/PolyMC/pull/1069. To address some of the comments in the discussion there:

> @Scrumplex: To avoid code duplication, instead of creating a CustomYggdrasil account type, maybe you could refactor Mojang accounts to support setting a custom URL instead. Then we can just extend our new, let's call it Yggdrasil account type with pre-defined auth servers and call that Mojang.

I decided for now to leave `CustomYggdrasil` as its own `AccountType` for reasons I outline here: https://github.com/PolyMC/PolyMC/pull/1069#issuecomment-1271078398. Willing to discuss and revisit this however.

> @fn2006: Just a heads up for if you do merge this, the JVM arguments that it uses to change the auth server were added in 1.16 which means that it'll function like an offline account if you use it in any version older than 1.16

Thanks, I didn't know support for these flags was added so recently. I plan on making a Fabric mod to backport the flags to older versions. Would it be appropriate to display a (disableable) warning message when launching an older version with a custom Yggdrasil account?

Original text of the pull request is duplicated below:

<hr>

This patch adds a new account type, "Custom Yggdrasil", which internally works much like a Mojang account but lets you specify your own third-party API servers that implement a Yggdrasil-like scheme.

When I refer to "API servers" I specifically mean four hosts:

- the "authentication" server, authserver.mojang.com
- the "account" server, api.mojang.com
- the "session" server, sessionserver.mojang.com
- the "services" server, api.minecraftservices.com

Vanilla Minecraft has built-in support for overriding the API servers at launch without any modifications to the game, which this patch takes advantage of---both the client and server will use the servers specified by the system properties `minecraft.api.*.host`. For example, running the vanilla server, you can specify your own API servers with:

```
java -Xmx1024M -Xms1024M \
	-Dminecraft.api.env=custom \
	-Dminecraft.api.auth.host=https://authserver.example.com \
	-Dminecraft.api.account.host=https://accountserver.example.com \
	-Dminecraft.api.session.host=https://sessionserver.example.com \
	-Dminecraft.api.services.host=https://servicesserver.example.com \
	-jar server.jar nogui
```

Here's an example of an Yggdrasil-compatible API server: https://github.com/rutkai/Mjolnir-Authentication-Server. I'm working on my own as well.

There are lots of reasons players might want to use API servers other than Mojang's. Your community might want to:
- Experience the benefits of offline servers without losing skins, encrypted connections, or account security
- Create additional accounts to allow users to have multiple players logged in to a server simultaneously
	- Have a "camera account" spectate your character for the purpose of content creation
	- Summon [bots](https://mineflayer.prismarine.js.org) to automate tedious tasks like AFKing, crafting, or even building map art. The technical community seems to have settled on Carpet Mod bots for some of these purposes, but if you ran your own authentication server, you could host a fleet of bot-controlled clients without modifying the server at all.
- Play Minecraft in a country or on a network where Mojang's servers are not accessible
- Create your own set of [miniblocks](http://minecraftplayerheadsdatabase.weebly.com/mini-blocks.html)
- Have skins with transparency, or high-resolution skins (would require a modded client as well)
- Have custom capes
- Serve access tokens that last longer, so you don't run into "Invalid session" errors as often. These errors can be annoying with modpacks that take a long time to restart.
- Opt out of Mojang's telemetry and/or chat reporting
- Have a backup authentication system in case Mojang's servers go down

I understand this proposal may be met with some controversy, mainly since it could be construed as enabling piracy. I hope I've given some valid reasons that legitimate, paying customers might want this feature, but I understand this project needs to stay well within the limits of Microsoft's EULA. So I've duplicated the logic from the "offline" account type so that a user can't add a "Custom Yggdrasil" account without first adding a Mojang or MSA account that owns Minecraft. With this DRM check in place, you aren't able to use a custom authentication server to make PolyMC into a "cracked" client.

Last, I'm not advocating that people migrate away from Mojang's infrastructure with this patch. I believe a centralized authentication system is one of Minecraft's greatest strengths---with one account you can play on almost any server. But for people who want to use their own API servers for whatever reason, the method should be as straightforward as possible, ideally baked in to everyone's favorite launcher.

evan-goode committed:
'Custom Yggdrasil' AccountType

Signed-off-by: Evan Goode <mail@evangoo.de>

pog5 reviewed (approved):
lgtm

Scrumplex commented:
I am still very unsure about this feature. I don't think there is any precedent of a reputable third party launcher implementing something this before.

Furthermore, I also wouldn't feel comfortable with Prism Launcher being the first here, as the project has already had enough drama.

evan-goode commented:
I understand where you're coming from, and I know interacting with nonfree software like Minecraft can be... dramatic, at times. I would hate to see this project be harassed as [youtube-dl has](https://www.eff.org/deeplinks/2020/11/github-reinstates-youtube-dl-after-riaas-abuse-dmca), for example. But in this case I think the risk to the project is quite negligible, and it's certainly no more risky than supporting offline accounts.

The relevant section of [Minecraft's EULA](https://www.minecraft.net/en-us/eula) is (bold mine):

> The one major rule is that you must not distribute anything we've made unless we specifically agree to it. By "distribute anything we've made" what we mean is:
> - give copies of our Game to anyone else;
> - make commercial use of anything we've made;
> - try to make money from anything we've made; or
> - **let other people get access to anything we've made in a way that is unfair or unreasonable;**

The DRM lock that Prism Launcher enforces on offline accounts, and custom Yggdrail accounts, should this PR be merged, places us solidly on the side of respecting this rule. Like other "reputable" launchers, including Mojang's own, we are not letting people get access to the Minecraft client unless we verify that they own the game.

I'd like to point out that this bit of the EULA is only relevant to the maintainers of Prism Launcher, not its users. The users would be paying Minecraft customers running an unmodified copy of the Minecraft client, customizing its behavior using command-line flags that the client expects and supports.

I don't know whether making philosophical arguments would be productive, but I firmly believe people should be able to play the game however they want after they've bought it, including playing it without with Mojang's servers. I bet this sentiment is shared by most members of this community. Fortunately, Minecraft's EULA is relatively generous; I see nothing in it that would prohibit either (1) the Prism maintainers from making this change or (2) the users from using this new feature.

> I don't think there is any precedent of a reputable third party launcher implementing something this before.

I see two reasons to explain why this hasn't been done before, and I don't think legal pressure is one of them:

1. Among people who bought the game, custom authentication servers are pretty niche topic. The people who will use this feature will likely be technical players or particularly privacy/freedom conscious users. There are probably some users who own the game, but currently have to use a less-reputable, possibly nonfree launcher to play with their friends who don't own the game, on servers that use custom authentication systems. The controversial chat reporting and mandatory telemetry introduced in 1.19 have sparked increased interest in alternative API servers, but the changes have to happen here first, in the launchers.

2. Until 1.16, custom API servers required both launcher support and a clientside mod like [authlib-injector](https://github.com/yushijinhun/authlib-injector). Now, the situation is different; both the client and server support overriding the authentication servers with command-line arguments, and launchers can implement support for third-party servers much more simply and elegantly.

ztgasdf commented:
> 2. Until 1.16, custom API servers required both launcher support and a clientside mod like [authlib-injector](https://github.com/yushijinhun/authlib-injector). Now, the situation is different; both the client and server support overriding the authentication servers with command-line arguments, and launchers can implement support for third-party servers much more simply and elegantly.

Could it be possible for you to add support for <1.16 versions using authlib-injector? Of course, it would be most ideal to have support for all versions, but it would at least increase the version support from 1.16+ to 1.7.6+.[1]

[1] - https://github.com/tinytengu/minecraft-authlib#what-is-up-with-this-particular-version

evan-goode commented:
> > 2. Until 1.16, custom API servers required both launcher support and a clientside mod like [authlib-injector](https://github.com/yushijinhun/authlib-injector). Now, the situation is different; both the client and server support overriding the authentication servers with command-line arguments, and launchers can implement support for third-party servers much more simply and elegantly.
> 
> Could it be possible for you to add support for <1.16 versions using authlib-injector? Of course, it would be most ideal to have support for all versions, but it would at least increase the version support from 1.16+ to 1.7.6+.[1]
> 
> [1] - https://github.com/tinytengu/minecraft-authlib#what-is-up-with-this-particular-version

Agreed that supporting older versions is very important. IMO the best way to support older versions (and skins too) is via a Fabric/Forge mod that the user installs on a particular instance.

- I'm not sure the launcher should make any automatic modifications to the game depending on which account is logged in. Any modifications to an instance should be made explicitly by the user. Maybe a dialog box prompting the user to install the mod, and/or a dedicated button in the "Mods" tab would be ideal.

- authlib-injector itself is not a great option. It expects a slightly different API than the Notchian client, see https://github.com/elyby/accounts/issues/2. It runs a local web server to proxy API requests for some reason. I also think it's strange that it's implemented as a Java agent rather than a traditional Fabric or Forge mod. Overall I think it's quite complex for something that can be done with a tiny Fabric mod.
- The approach of https://github.com/tinytengu/minecraft-authlib is interesting in that it doesn't require a Java agent or mod loader. But Prism cannot distribute a patched authlib ourselves; per Mojang's EULA, we can only distribute mods. We could force the client to use the newer version of Mojang's authlib that supports custom API servers, but skins won't work since it'll still check signatures against Mojang's public key, and so the user would likely want to fix that with a mod anyway.
 
Given the info in the README here https://github.com/tinytengu/minecraft-authlib, authlib has changed so little over the years that it should be pretty easy to maintain Fabric and Forge mods compatible with at least 1.7.6+.

evan-goode committed:
Allow multiple accounts with the same player UUID

Signed-off-by: Evan Goode <mail@evangoo.de>

TheKodeToad commented:
I managed to get it working on older versions without any modifications - just updating the authlib version. It works fine with Java 8 and no other lib changes with authlib version 2.1.28.

Something similar to #443 could be used to redirect urls though.

x0rp01s0n commented:
After 1.19 update privatifying the auth system came in as only solution for me,
heres a custom launcher that supports a custom account, https://github.com/MrLetsplay2003/ShittyAuthLauncher ,
technically the game already got taken away from microsoft they took it too far xd

TheKodeToad commented:
Huh? There's the -Dminecraft.thing=ithjawkj,mz properties.

tesseract-two commented:
Hello, any updates on this? I would use this because I:
- Have a Mojang Account
- Run my own server
- Appreciate privacy
- Would use the slightly better security (potential 2FA, can provision and control accounts for my players, someone's Mojang token being stolen doesn't effect my server)
- Do some roleplay (Multiple accounts per person, can have any name, change it whenever I want, and have it show up in book signing and commands)

This seems like the best solution to continue playing on my server without migrating. Thank you all for your hard work on this awesome launcher, and have a good day. :)

Krysztal112233 commented:
Great job, it's very necessary for people who don't want to use Mojang and Microsoft. But I think it would be better to have an alert to warn users about the dangers of using third-party Yggdrasil services. @evan-goode

evan-goode commented:
Thanks for your review, @Krysztal112233. I think most of the people using this feature will know what they're doing, but that's a valid point; there is the risk that someone may not understand that their credentials will be sent to the authentication server they specify, and they may mistakenly send their Mojang/Microsoft username and password to a third party. Maybe instead of adding a separate warning dialog, the message on the "Add Account" dialog could be changed to something like:

> Please enter your email/username, password, and the URLs of your API servers.
> Caution! Your credentials will be sent to the authentication server you specify.

Are there any other dangers you can think of? Again, I don't think there is any interpretation of the Mojang EULA that would imply users are not allowed to use custom API servers, so there is no issue there.

Perhaps this message should also link to a page in the Prism Launcher documentation about what the API servers are, and how to modify the game to get custom API servers working with older versions of Minecraft.

evan-goode commented:
Rebased and added a missing signed-off-by.

Krysztal112233 commented:
Bravo! That's what I meant; other than that I don't think there should be too many problems, and really hope this PR gets merged: with launchers like HMCL, they've had similar functionality for a long time.
________________________________
发件人: Evan Goode ***@***.***>
发送时间: 2023年5月13日 23:23
收件人: PrismLauncher/PrismLauncher ***@***.***>
抄送: Krysztal112233 ***@***.***>; Mention ***@***.***>
主题: Re: [PrismLauncher/PrismLauncher] 'Custom Yggdrasil' AccountType (PR #543)


Thanks for your review, @Krysztal112233<https://github.com/Krysztal112233>. I think most of the people using this feature will know what they're doing, but that's a valid point; there is the risk that someone may not understand that their credentials will be sent to the authentication server they specify, and they may mistakenly send their Mojang/Microsoft username and password to a third party. Maybe instead of adding a separate warning dialog, the message on the "Add Account" dialog could be changed to something like:

Please enter your email/username, password, and the URLs of your API servers.
Caution! Your credentials will be sent to the authentication server you specify.

Are there any other dangers you can think of? Again, I don't think there is any interpretation of the Mojang EULA that would imply users are not allowed to use custom API servers, so there is no issue there.

Perhaps this message should also link to a page in the Prism Launcher documentation about what the API servers are, and how to modify the game to get custom API servers working with older versions of Minecraft.

―
Reply to this email directly, view it on GitHub<https://github.com/PrismLauncher/PrismLauncher/pull/543#issuecomment-1546688804>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AE4P55I4KTEWG6FENDYBHNLXF6RNPANCNFSM6AAAAAASS3XRFM>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

Krysztal112233 commented:
Can you check this pull request again? @Scrumplex  Regarding what you are wondering, there is a third-party launcher called [HMCL](https://github.com/huanghongxun/HMCL) that has been implementing this feature for at least two years and has a very high level of trust in some regions: since this feature is actually a Yggdrasil server built by the third-party server itself, its owner repeatedly warns the player to always separate from the Microsoft account, and any accidents that arise are will not be covered by the launcher. 

This is not really a dramatic content, it is very reasonable for some needs.

For this pull request, at least in China, is the demand of tens of thousands of players: for some reason, third-party servers with third-party logins are actually very popular, and there is no news of theft of Microsoft accounts - because players who use these services are often warned by some people such as the owners of the third-party services.

These words may seem a little rude, but they are the truth, so please forgive my poor English vocabulary if it makes you feel uncomfortable.

evan-goode committed:
Use correct services server URL for SkinDelete

Signed-off-by: Evan Goode <mail@evangoo.de>

evan-goode committed:
Correctly use CustomYggdrasilRefresh, add warning message

Signed-off-by: Evan Goode <mail@evangoo.de>

evan-goode commented:
I added a warning message to the dialog where you add a Custom Yggdrasil account. It now reads:

> Please enter your username (sometimes an email address), password, and the URLs of your API servers.
> 
> **Caution!** Your username and password will be sent to the authentication server you specify!

I also fixed up a couple things I missed:
- The correct method is now used for refreshing Custom Yggdrasil accounts; the MSA migration check should not be performed
- The correct services server URL for the account will be used for the SkinDelete task.

osfanbuff63 commented:
Edit: Github Mobile prevented me from seeing the whole description :skull:

<details>
<summary>Click to view old text</summary>

Part of me thinks this could easily become a feature for piracy, which Prism Launcher should have no part in. Are there limitations that prevent this? If not, this could easily get Prism removed from Modrinth's recommendations, among others - and you'd probably lose a good number of users

</details>

evan-goode commented:
> Part of me thinks this could easily become a feature for piracy, which Prism Launcher should have no part in. Are there limitations that prevent this? If not, this could easily get Prism removed from Modrinth's recommendations, among others - and you'd probably lose a good number of users

Yes, I've discussed this issue above but I can elaborate. In this patch, I include the same anti-piracy protection that is already used in Prism Launcher for offline accounts: if you try to add a "Custom Yggdrasil" account without first adding a valid Microsoft or Mojang account, Prism Launcher will refuse and show the error message, "You must add a Microsoft or Mojang account that owns Minecraft before you can add an account on a custom authentication server". The message is implemented here: https://github.com/PrismLauncher/PrismLauncher/blob/984791c30aa2de13253544a7e7d3b0fa9c767f3a/launcher/ui/pages/global/AccountListPage.cpp#L161

There was some discussion on the [original PolyMC PR](https://github.com/PolyMC/PolyMC/pull/1069) about making this anti-piracy check more strict, but I think what we have is enough and stricter measures would be impractical and ineffective. If this patch is merged, tightening the DRM restrictions on both offline accounts and Custom Yggdrasil accounts could be done in a later PR.

Aside from the technical details of the DRM itself, yes, there is the broader possibility of Prism Launcher merging this and becoming more associated with communities which are indifferent to piracy, thereby losing respect among more mainstream (notably, more Western) communities.

But again, with the DRM check, I don't think there is significant risk to Prism Launcher's credibility. People who don't own Minecraft will continue using some Prism Launcher fork, HMCL, or some other launcher. The audience for this patch will overwhelmingly comprise people with Minecraft licenses. According to @Krysztal112233, third-party authentication systems are very popular in China, even among people with valid Minecraft accounts (perhaps something to do with the Great Firewall and/or [China Edition](https://minecraft.fandom.com/wiki/China_Edition)?) As I, @tesseract-two, and others have expressed in this thread, there are freedom and privacy-related reasons why paying Minecraft users would want to use a custom authentication system.

Speaking of Modrinth, they [currently recommend](https://docs.modrinth.com/docs/modpacks/playing_modpacks/) a Chinese launcher called BakaXL in addition to ATLauncher, MultiMC, and Prism Launcher. BakaXL supports custom Yggdrasil accounts via authlib-injector, and does not enforce any anti-piracy checks. You just have to click "I don't have the game" when setting up the launcher and you can sign in with a third-party auth provider. At least in Modrinth's case, I think Prism Launcher would have nothing to worry about. 

Ultimately this debate is as much political as it is technical. It is in Microsoft's interests to enforce DRM on the game by centrally hosting the API servers, and it is in the interests of Prism Launcher's users to be free to use their own API servers. Maybe the maintainers could put it to a vote or something.

evan-goode commented:
Update on BakaXL: the QuiltMC project had a good, lengthy discussion about whether to endorse it here: https://github.com/QuiltMC/quiltmc.org/issues/51. They decided to keep it on the QuiltMC page, and Modrinth then followed their lead: https://github.com/modrinth/docs/pull/69.

According to that thread, BakaXL actually does have some kind of region-locking DRM. Outside of China, they claim to use a strategy similar to Prism Launcher for offline accounts (but possibly not custom Yggdrasil accounts, it's unclear), and this DRM was key in QuiltMC's decision.

FWIW their DRM does not seem to work very well, I have a United States IP and I was able to launch Minecraft without signing in to a Mojang account on BakaXL version 3.2.5.0.

osfanbuff63 commented:
In reply to the first comment after mine:

Alright, that makes more sense now. I don't think this needs to be more strict than that.

Big-Cake-jpg commented:
> FWIW their DRM does not seem to work very well, I have a United States IP and I was able to launch Minecraft without signing in to a Mojang account on BakaXL version 3.2.5.0.

(Google Translate)

I once asked the developers in the BakaXL QQ group that BakaXL's copyright protection mechanism determines whether to open the offline login option by detecting the country and region settings of Windows.

I'm trying to talk to the developers and let them decide whether to add another layer of copyright protection based on IP detection in BakaXL.

---

(Chinese)

我曾询问过 BakaXL QQ 群中的开发者，BakaXL 的版权保护机制是通过检测 Windows 的国家和地区设置来决定是否开放离线登录选项的。

我正在尝试和开发者进行沟通，让他们决定是否在 BakaXL 中再添加一层基于 IP 检测的版权保护机制。

TT702 commented:
> Update on BakaXL: the QuiltMC project had a good, lengthy discussion about whether to endorse it here: https://github.com/QuiltMC/quiltmc.org/issues/51. They decided to keep it on the QuiltMC page, and Modrinth then followed their lead: https://github.com/modrinth/docs/pull/69.
> 
> 
> 
> According to that thread, BakaXL actually does have some kind of region-locking DRM. Outside of China, they claim to use a strategy similar to Prism Launcher for offline accounts (but possibly not custom Yggdrasil accounts, it's unclear), and this DRM was key in QuiltMC's decision.
> 
> 
> 
> FWIW their DRM does not seem to work very well, I have a United States IP and I was able to launch Minecraft without signing in to a Mojang account on BakaXL version 3.2.5.0.

Hi, I'm the developer of BakaXL Launcher, I'm sorry to hear that Copyright Protection function was not work properly on your computer, this was not supposed to happen.

Because I didn't receive any report on this function was failed before so I would like to ask you to help me solve this issue.

Please open an issue and provide the latest log file that generate by BakaXL to BakaXL's GitHub repository if possible, much appreciated!

tesseract-two commented:
Would any of the maintainers mind taking another look at this? You already support offline accounts, what additional dangers does this present? This is also a official feature, and requires more setup than turning offline mode on in your server.properties.

I'd still use this after migrating because I'd prefer MS not know what servers I'm playing on, when, and with who. Offline mode doesn't provide any net privacy benefit for me because it disables encryption (imo a bad idea in 2023). And of course the silly roleplay benefits.

Thank you for your time working on this project!

TheKodeToad commented:
I personally think this is a pretty nice idea. Yggdrasil is pretty much dead, so why not allow people to replace it (edit: I was missing some aspects of this)? Additionally, you can pretty much already do this. Just add the flags, change the token (afaik). You can even do it in Mojang's official launcher (again, afaik, in theory)! This just integrates it in a sane way.

Also, I agree, offline mode just isn't secure.

Of course, I'm only one maintainer (and a very new one) so we'd have to agree on this

Edit: I wasn't very active on the PolyMC repo but seems DioEgizio downvoted it... so already two other people not sure about this. :woman_shrugging: but it's pretty much taking advantage of an official feature.

TheKodeToad commented on launcher/minecraft/auth/AccountData.cpp:
```diff
@@ -473,6 +527,9 @@ QString AccountData::accountDisplayString() const {
         }
+        case AccountType::CustomYggdrasil: {
+            return userName();
+        }
```
```suggestion
        case AccountType::Mojang:
        case AccountType::CustomYggdrasil: {
            return userName();
        }
```
the braces do kind of look weird here tbh :woman_shrugging:

evan-goode replied:
Done!

TheKodeToad commented on launcher/minecraft/auth/AccountList.cpp:
```diff
@@ -122,8 +122,17 @@ void AccountList::addAccount(const MinecraftAccountPtr account)
+                existingAccount = i;
+                break;
+            }
+        }
```
I think this could use std::find. I know this was here before *sigh* but this feels like a bit of a code smell tbh

evan-goode replied:
Sure, I've replaced this with a `std::find_if`. Might as well use it for new code. It might be more readable, though C++ does its best to make that not the case :)

TheKodeToad replied:
oh, thank you! I got a little confused

TheKodeToad commented on launcher/minecraft/auth/MinecraftAccount.cpp:
```diff
@@ -82,6 +83,26 @@ MinecraftAccountPtr MinecraftAccount::createFromUsername(const QString &username
+    auto account = makeShared<MinecraftAccount>();
+    account->data.type = AccountType::CustomYggdrasil;
+    account->data.yggdrasilToken.extra["userName"] = username;
+    account->data.yggdrasilToken.extra["clientToken"] = QUuid::createUuid().toString().remove(QRegularExpression("[{}-]"));
```
```suggestion
    account->data.yggdrasilToken.extra["clientToken"] = QUuid::createUuid().toString(QUuid::Id128);
```

evan-goode replied:
Done, I've replaced the other `.remove(QRegularExpression(...))` in the file as well.

TheKodeToad commented on launcher/minecraft/auth/MinecraftAccount.cpp:
```diff
@@ -140,6 +161,17 @@ shared_qobject_ptr<AccountTask> MinecraftAccount::login(QString password) {
+    Q_ASSERT(m_currentTask.get() == nullptr);
+
+    m_currentTask.reset(new CustomYggdrasilLogin(&data, password));
+    connect(m_currentTask.get(), SIGNAL(succeeded()), SLOT(authSucceeded()));
+    connect(m_currentTask.get(), SIGNAL(failed(QString)), SLOT(authFailed(QString)));
```
```suggestion
    connect(m_currentTask.get(), &Task::succeeded, this, &MinecraftAccount::authSucceeded));
    connect(m_currentTask.get(), &Task::failed, this, &MinecraftAccount::authFailed));
```
only if you want

evan-goode replied:
Done

TheKodeToad commented on launcher/minecraft/auth/Yggdrasil.cpp:
```diff
@@ -72,6 +73,8 @@ void Yggdrasil::refresh() {
     req.insert("clientToken", m_data->clientToken());
     req.insert("accessToken", m_data->accessToken());
+
+    qDebug() << "refreshing, access token is" << m_data->accessToken();
```
*ouch*
```suggestion
    qCDebug(authCredentials()) << "refreshing, access token is" << m_data->accessToken();
```

evan-goode replied:
Whoops. There's always a `print` you miss. I'll just take this out instead of sanitizing it.

TheKodeToad commented on launcher/minecraft/auth/steps/MinecraftProfileStep.cpp:
```diff
@@ -44,7 +44,7 @@ void MinecraftProfileStep::onRequestDone(
     if (error == QNetworkReply::ContentNotFoundError) {
         // NOTE: Succeed even if we do not have a profile. This is a valid account state.
-        if(m_data->type == AccountType::Mojang) {
+        if(m_data->type == AccountType::Mojang || m_data->type == AccountType::CustomYggdrasil) {
```
Custom servers cannot dictate whether you own the game imo
I don't know a lot about the codebase, but this may affect the checks of whether you own the game.

evan-goode replied:
I agree, I've changed it to grant `canPlayMinecraft` and `ownsMinecraft` when the account is created, like we do for offline accounts, instead of doing it here.

TheKodeToad commented on launcher/minecraft/auth/steps/MinecraftProfileStep.cpp:
```diff
@@ -87,7 +87,7 @@ void MinecraftProfileStep::onRequestDone(
     }
 
-    if(m_data->type == AccountType::Mojang) {
+    if(m_data->type == AccountType::Mojang || m_data->type == AccountType::CustomYggdrasil) {
```
^

TheKodeToad commented on launcher/minecraft/auth/steps/MinecraftProfileStepMojang.cpp:
```diff
@@ -47,7 +48,7 @@ void MinecraftProfileStepMojang::onRequestDone(
     if (error == QNetworkReply::ContentNotFoundError) {
         // NOTE: Succeed even if we do not have a profile. This is a valid account state.
-        if(m_data->type == AccountType::Mojang) {
+        if(m_data->type == AccountType::Mojang || m_data->type == AccountType::CustomYggdrasil) {
```
^

TheKodeToad commented on launcher/minecraft/auth/steps/MinecraftProfileStepMojang.cpp:
```diff
@@ -90,7 +91,7 @@ void MinecraftProfileStepMojang::onRequestDone(
     }
 
-    if(m_data->type == AccountType::Mojang) {
+    if(m_data->type == AccountType::Mojang || m_data->type == AccountType::CustomYggdrasil) {
```
^ (why is there so much duplication :sob:)

TheKodeToad commented on launcher/minecraft/services/SkinUpload.h:
```diff
@@ -18,13 +19,13 @@ class SkinUpload : public Task
     Model m_model;
     QByteArray m_skin;
-    QString m_token;
+    MinecraftAccountPtr m_acct;
```
`m_account` is more readable!

evan-goode replied:
Agree, I've changed `m_acct` to `m_account` everywhere. There weren't too many other places outside the scope of this PR.

TheKodeToad commented on launcher/ui/dialogs/CustomYggdrasilLoginDialog.cpp:
I had some suggestions but turns out you're not to blame!

TheKodeToad reviewed (changes requested):
For now, here are some changes i think'd be nice
Also, I think it should have an extra confirmation dialog, *especially* if you're using http, but it'd be nice for any url in case of accidents

evan-goode committed:
Custom Yggdrasil: Readability fixes

Also made the MinecraftEntitlement for Custom Yggdrasil accounts work
just like offline accounts---instead of checking the reply from the auth
server, Custom Yggdrasil accounts are granted canPlayMinecraft and
ownsMinecraft when they are created, in MinecraftAccount.cpp.

Signed-off-by: Evan Goode <mail@evangoo.de>

evan-goode commented:
> For now, here are some changes i think'd be nice

Thanks for the review!

>Also, I think it should have an extra confirmation dialog, especially if you're using http, but it'd be nice for any url in case of accidents

Yeah, like a warning that pops up after you hit "OK" on the Add Account dialog? Something like:

> Warning! You are about to send the username and password you entered to an unofficial, third-party authentication server! Never use your Mojang or Microsoft password for a third-party account! 

with "Proceed" and "Cancel" buttons?

I agree that the warning should show regardless of whether the URL is http:// or https://.

These kinds of things also tend to have a 3, 2, 1 countdown before you can continue, but I don't know whether that would be worth the implementation cost and annoyance.

Since you have to enter four separate URLs, it's unlikely someone will create a Custom Yggdrasil account accidentally, but I guess the threat model is someone saying, "hey kid, you can get a cool free cape if you go to 'Custom Yggdrasil' and put in this server...".

ghost commented:
@evan-goode
Different servers have online-mode set to true for different skin servers. For example, one server might have you sign up on LittleSkin and another on Blessing Skin in order to login through it with authlib-injector. On HMCL you can add these authenticators manually by setting the server to https://littleskin.cn/api/yggdrasil/ or https://skin.prinzeugen.net/api/yggdrasil/ instead.

I compiled your fork, but it asks for way more: four different URLs. The Yggrasil API seems to have a single URL that should work in their place. I don't get how to use it for any Yggdrasil server because of it.

Info: https://github.com/yushijinhun/authlib-injector/wiki/Yggdrasil-%E6%9C%8D%E5%8A%A1%E7%AB%AF%E6%8A%80%E6%9C%AF%E8%A7%84%E8%8C%83

evan-goode commented:
@poopingpenis yes, thanks for bringing that up. The vanilla authlib uses four different API servers. Authlib-injector remaps each of these to be under one API root here: https://github.com/yushijinhun/authlib-injector/blob/b68049208a63b974f2cea34d8c8ab3770a9e110c/src/main/java/moe/yushi/authlibinjector/httpd/DefaultURLRedirector.java#L36.

So if your authlib-injector-compatible server is https://example.com, then the mapping becomes:
- authserver.mojang.com <=> https://example.com/authserver
- api.mojang.com <=> https://example.com/api
- sessionserver.mojang.com <=> https://example.com/sessionserver
- api.minecraftservices.com <=> https://example.com/minecraftservices

Authlib-injector's API and Mojang's API (documented [here](https://c4k3.github.io/wiki.vg/Authentication.html), [here](https://c4k3.github.io/wiki.vg/Protocol_Encryption), and [here](https://wiki.vg/Mojang_API)) are close enough that you should be able to use authlib-injector servers with this patch. I can use a Blessing Skin account on Prism Launcher with the following servers:
- auth server: https://skin.prinzeugen.net/api/yggdrasil/authserver
- account server: https://skin.prinzeugen.net/api/yggdrasil/api
- session server: https://skin.prinzeugen.net/api/yggdrasil/sessionserver
- services server: https://skin.prinzeugen.net/api/yggdrasil/minecraftservices

Unfortunately, beyond simple authentication, Mojang's API and authlib-injector's API start to diverge. For example, you change your skin on the Mojang API using https://example.com/minecraft/profile/skins, but using authlib-injector's API it's https://example.com/api/user/profile/{uuid}/skin. So changing/deleting your skin does not work with Prism + Blessing Skin. There are probably historical reasons why authlib-injector doesn't always follow Mojang's API, maybe authlib-injector did some things first.

The changes I am proposing stay true to the way the vanilla game expects custom API servers to work, not the way authlib-injector does it. As you noticed, there is certainly a usability problem since most API servers currently work with authlib-injector. Some options for reconciling the differences are:

- My preferred option: authlib-injector-compatible authentication servers could add support for the missing routes from Mojang's official API to fix skins and other functionality. They could add a "Prism Launcher" or "Vanilla authlib" section to their documentation that lists out the four URLs that the vanilla client and server expect, or alternatively, they could add a `/vanilla` route that puts all the functionality under one endpoint so you can use the same URL four times in Prism Launcher. Authlib-injector could add the missing Mojang routes to its API documentation so new auth servers know to implement these.
- Prism Launcher could add an "authlib-injector-style" option to the Custom Yggdrasil account page, where you enter one URL instead of four, and we automatically append those /authserver, /api, /sessionserver, /minecraftservices paths. It's more user-friendly this way, but it's incongruent with vanilla authlib.
- Prism Launcher could abandon this PR and instead fully support authlib-injector, including supporting their skin-changing API routes, automatically launching the game with authlib-injector, and adding authlib-injector accounts with a single URL. I think there are already some forks that do it this way.

ghost commented:
@evan-goode 
Managed to login, but when I open a singleplayer world, I'm missing my skin. I ran a test server in online mode with authlib-injector, and sure enough I can enter the server, but still no skin. I tried entering again with an offline profile, and I couldn't join, so authentication itself is not the issue...

On my Prism Launcher profiles, it loads the head of my skin for its icon. But not in-game...

evan-goode commented:
@poopingpenis that's the main drawback of vanilla authlib: `com.mojang.authlib.yggdrasil.TextureUrlChecker` checks that texture URLs returned by the authentication server are on a Mojang domain. Some third-party authentication servers support forwarding textures from Mojang accounts; that would work with vanilla authlib. But if you want to host your own skins, you'll need to patch authlib somehow, either by using a Fabric/Forge mod, some kind of Java agent like authlib-injector, or replacing authlib with a custom version.

I don't yet have a definite idea of how best to fix skins, or whether Prism Launcher should try to automatically modify authlib when using a Custom Yggdrasil account.

ghost commented:
@evan-goode Custom skin support with authlib-injector would definitely be ideal since its extended skin and cape system is one of its benefits, and its support for them in closed networks (for example China or Cuba's SNET/Tinored).

I like how HMCL supports it, but the client itself is not to my taste so I look forward to the development of this feature, thanks

evan-goode committed:
Custom Yggdrasil: Add extra confirmation dialog

Signed-off-by: Evan Goode <mail@evangoo.de>

evan-goode commented:
@TheKodeToad I added that extra confirmation dialog.

TheKodeToad commented:
:blobfoxree: Ok then

TheKodeToad reviewed (approved):
Implementation seems good. However the chances of merging seem low.

TheKodeToad committed:
Merge branch 'develop' into custom-yggdrasil

Signed-off-by: TheKodeToad <TheKodeToad@proton.me>

OrzMiku commented:
As you can see, both HMCL, BakaXL, and PCL2 launchers support customizing Yggdrasil accounts through authlib-injector. Moreover, BakaXL has preserved this launcher in the documentation on Modrinth and QuiltMC, indicating a certain level of recognition among players. Additionally, the Prism Launcher itself has anti-piracy measures in place, so I believe it's entirely acceptable for legitimate paying players to use custom Yggdrasil accounts.

I really appreciate the feature in Prism Launcher that allows exporting Modpacks. However, the issue with 'Custom Yggdrasil' still remains unresolved. My server uses the "Union Authentication" from the "Minecraft University Alliance," which relies on the use of Authlib-Injector. Therefore, I am in great need of Prism Launcher's support in this regard at the moment.

dozed-dev commented:
Would be nice to have this. If i rebase this to current dev branch, is there any chance it'll get merged? I'm not familiar with prism codebase though :(

evan-goode commented:
> Would be nice to have this. If i rebase this to current dev branch, is there any chance it'll get merged? I'm not familiar with prism codebase though :(

I think it's time to close this PR, actually. It's been open (continued from https://github.com/PolyMC/PolyMC/pull/1069) for almost two years, and though it's generally received good feedback, the maintainers don't seem to want it.

I've continued work on supporting alternative auth servers in my (soft) fork of Prism Launcher, Fjord Launcher. Fjord Launcher uses the code in this PR and additionally supports authlib-injector, which allows skins to work and lets us support client versions older than 1.16. My idea to replace authlib-injector with Fabric mods would probably not be worth it, though I would still like to replace authlib-injector with something more lightweight in the future. Fjord distributes authlib-injector [via](https://github.com/unmojang/meta/blob/6671846cb7fbd4c5ca303d3b27176128b259f968/updateAuthlibInjector.py) a fork of Prism's metadata server. The launcher shows a (disableable) prompt to install authlib-injector on an instance when it's needed, in a dialog box similar to the "Install Loader" dialog. This approach enables a pretty seamless UX while still letting the user control whether to use authlib-injector, and what version to use.

If the maintainers ever come to agreement, I would be happy to keep working on upstreaming this, but I would rather do that in a new PR based on the latest changes in Fjord Launcher. You can contact me on Matrix, @ in bio.

Thank you to everyone who weighed in and gave feedback here!
